google/cloud/bigquery/magics/line_arg_parser/lexer.py

```diff
@@ -0,0 +1,278 @@
+    PARSE_NON_PARAMS_OPTIONS = ()  # parsing options other than "--params"
+    PARSE_PARAMS_OPTION = ()  # parsing the "--params" option
+    STATE_END = ()
+
+
+class Lexer(object):
+    """Lexical analyzer for tokenizing the cell magic input line."""
+
+    _GRAND_PATTERNS = {
+        LexerState.PARSE_POS_ARGS: re.compile(
+            "|".join(
+                itertools.chain(
+                    token_types["state_parse_pos_args"].values(),
+                    token_types["common"].values(),
+                )
+            )
+        ),
+        LexerState.PARSE_NON_PARAMS_OPTIONS: re.compile(
+            "|".join(
+                itertools.chain(
+                    token_types["state_parse_non_params_options"].values(),
+                    token_types["common"].values(),
+                )
+            )
+        ),
+        LexerState.PARSE_PARAMS_OPTION: re.compile(
+            "|".join(
+                itertools.chain(
+                    token_types["state_parse_params_option"].values(),
+                    token_types["common"].values(),
+                )
+            )
+        ),
+    }
+
+    def __init__(self, input_text):
+        self._text = input_text
+
+    def __iter__(self):
+        # Since re.scanner does not seem to support manipulating inner scanner states,
+        # we need to implement lexer state transitions manually using special
+        # non-capturing lookahead token patterns to signal when a state transition
+        # should be made.
+        # Since we don't have "nested" states, we don't really need a stack and
+        # this simple mechanism is sufficient.
+        state = LexerState.PARSE_POS_ARGS
+        offset = 0  # the number of characters processed so far
+
+        while state != LexerState.STATE_END:
+            token_generator = self._get_state_token_generator(state, offset)
+
+            for maybe_token in token_generator:  # pragma: NO COVER
+                if isinstance(maybe_token, StateTransition):
+                    state = maybe_token.new_state
+                    offset = maybe_token.total_offset
+                    break
+
+                if maybe_token.type_ != TokenType.WS:
+                    yield maybe_token
+
+                if maybe_token.type_ == TokenType.EOL:
+                    state = LexerState.STATE_END
+                    break
+
+    def _get_state_token_generator(self, state, current_offset):
+        """Return token generator for the current state starting at ``current_offset``.
+
+        Args:
+            state (LexerState): The current lexer state.
+            current_offset (int): The offset in the input text, i.e. the number
+                of characters already scanned so far.
+
+        Returns:
+            A generator yielding ``Token`` and ``StateTransition`` instances.
+        """
+        pattern = self._GRAND_PATTERNS[state]
+        scanner = pattern.scanner(self._text, current_offset)
```

Comment:
It's something that has been around for many years, and yes, it's less known due to not being documented. The `finditer()` method thus seems better, even if just to avoid any confusion.